api/ngrok/v1alpha1/agentendpoint_types.go

```diff
@@ -0,0 +1,169 @@
+// +kubebuilder:resource:shortName=aep
+// +kubebuilder:printcolumn:name="URL",type="string",JSONPath=".status.assignedURL"
+// +kubebuilder:printcolumn:name="Traffic Policy",type="string",JSONPath=".status.trafficPolicy"
+// +kubebuilder:printcolumn:name="Bindings",type="string",JSONPath=".spec.bindings"
+// +kubebuilder:printcolumn:name="Status",type="string",JSONPath=".status.conditions[?(@.type=='Status')].status"
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+type AgentEndpoint struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   AgentEndpointSpec   `json:"spec,omitempty"`
+	Status AgentEndpointStatus `json:"status,omitempty"`
+}
+
+// AgentEndpointSpec defines the desired state of an AgentEndpoint
+type AgentEndpointSpec struct {
+	// The unique URL for this agent endpoint. This URL is the public address. The following formats are accepted
+	// Domain - example.org
+	//     When using the domain format you are only defining the domain. The scheme and port will be inferred.
+	// Origin - https://example.ngrok.app or https://example.ngrok.app:443 or tcp://1.tcp.ngrok.io:12345 or tls://example.ngrok.app
+	//     When using the origin format you are defining the protocol, domain and port. HTTP endpoints accept ports 80 or 443 with respective protocol.
+	// Scheme (shorthand) - https:// or tcp:// or tls:// or http://
+	//     When using scheme you are defining the protocol and will receive back a randomly assigned ngrok address.
+	// Empty - ``
+	//     When empty your endpoint will default to be https and receive back a randomly assigned ngrok address.
+	// Internal - some.domain.internal
+	//     When ending your url with .internal, an internal endpoint will be created. nternal Endpoints cannot be accessed directly, but rather
+	//     can only be accessed using the forward-internal traffic policy action.
+	//
+	// +kubebuilder:validation:Required
+	URL string `json:"url"`
+
+	// Allows configuring a TrafficPolicy to be used with this AgentEndpoint
+	//
+	// +kubebuilder:validation:Optional
+	TrafficPolicy *TrafficPolicyCfg `json:"trafficPolicy,omitempty"`
+
+	// Human-readable description of this agent endpoint
+	//
+	// +kubebuilder:default:=`Created by the ngrok-operator`
+	Description string `json:"description,omitempty"`
+
+	// String of arbitrary data associated with the object in the ngrok API/Dashboard
+	//
+	// +kubebuilder:default:=`{"owned-by":"ngrok-operator"}`
+	Metadata string `json:"metadata,omitempty"`
+
+	// List of Binding IDs to associate with the endpoint
+	// Accepted values are "public", "internal", or strings matching the pattern "k8s/*"
+	//
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Items=pattern=`^(public|internal|k8s/.*)$`
+	Bindings []string `json:"bindings,omitempty"`
+}
+
+type TrafficPolicyCfgType string
+
+const (
+	TrafficPolicyCfgType_K8sRef TrafficPolicyCfgType = "targetRef"
```

Comment:
We currently support k8s object refs and inline for the traffic policy, but we might want to support other ways to supply one in the future such as configuring one in the ngrok cloud and having one or more clusters running the operator reference that one and have the definition pulled from the cloud.